patches/tock/10-avoid-app-reentry.patch

```diff
@@ -0,0 +1,329 @@
+@@ -25,73 +25,15 @@ pub const CTAP_SUBSCRIBE_TRANSMIT_OR_RECEIVE: usize = 3;
+ pub const CTAP_CALLBACK_TRANSMITED: usize = 1;
+ pub const CTAP_CALLBACK_RECEIVED: usize = 2;
+ 
+-#[derive(Clone, Copy, PartialEq, Eq)]
+-enum Side {
+-    Transmit,
+-    Receive,
+-    TransmitOrReceive,
+-}
+-
+-impl Side {
+-    fn can_transmit(&self) -> bool {
+-        match self {
+-            Side::Transmit | Side::TransmitOrReceive => true,
+-            Side::Receive => false,
+-        }
+-    }
+-
+-    fn can_receive(&self) -> bool {
+-        match self {
+-            Side::Receive | Side::TransmitOrReceive => true,
+-            Side::Transmit => false,
+-        }
+-    }
+-}
+-
+-#[derive(Default)]
+-pub struct App {
+-    // Only one app can be connected to this driver, to avoid needing to route packets among apps.
+-    // This field tracks this status.
+-    connected: bool,
+-    // Currently enabled transaction side. Subscribing to a callback or allowing a buffer
+-    // automatically sets the corresponding side. Clearing both the callback and the buffer resets
+-    // the side to None.
+-    side: Option<Side>,
+-    callback: Option<Callback>,
+-    buffer: Option<AppSlice<Shared, u8>>,
+-    // Whether the app is waiting for the kernel signaling a packet transfer.
+-    waiting: bool,
+-}
+-
+-impl App {
+-    fn check_side(&mut self) {
+-        if self.callback.is_none() && self.buffer.is_none() && !self.waiting {
+-            self.side = None;
+-        }
+-    }
+-
+-    fn set_side(&mut self, side: Side) -> bool {
+-        match self.side {
+-            None => {
+-                self.side = Some(side);
+-                true
+-            }
+-            Some(app_side) => side == app_side,
+-        }
+-    }
+-
+-    fn is_ready_for_command(&self, side: Side) -> bool {
+-        self.buffer.is_some() && self.callback.is_some() && self.side == Some(side)
+-    }
+-}
+-
+ pub trait CtapUsbClient {
+     // Whether this client is ready to receive a packet. This must be checked before calling
+-    // packet_received().
+-    fn can_receive_packet(&self) -> bool;
++    // packet_received(). If App is not supplied, it will be found from the implemntation's
++    // members.
++    fn can_receive_packet(&self, app: &Option<&mut App>) -> bool;
+ 
+     // Signal to the client that a packet has been received.
+-    fn packet_received(&self, packet: &[u8; 64]);
++    // If App is not supplied, it will be found from the implemntation's members.
++    fn packet_received(&self, packet: &[u8; 64], app: Option<&mut App>);
+ 
+     // Signal to the client that a packet has been transmitted.
+     fn packet_transmitted(&self);
+@@ -106,38 +48,49 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> CtapUsbSyscallDriver<'a, 'b, C> {
+     pub fn new(usb_client: &'a ClientCtapHID<'a, 'b, C>, apps: Grant<App>) -> Self {
+         CtapUsbSyscallDriver { usb_client, apps }
+     }
++
++    fn _packet_received(&self, packet: &[u8; 64], app: &mut App) {
```

Comment:
Names starting with `_` usually indicate unused. The function doesn't look like it's unused, is this intended?

Reply:
That was me forgetting naming conventions in different languages.

Fixed.